src/contexts/providers.context.tsx

```diff
@@ -127,35 +145,59 @@ const EthereumProviderProvider: FC = (props) => {
       window.location.reload();
     };
 
-    if (internalProvider && "on" in internalProvider && typeof internalProvider.on === "function") {
-      internalProvider.on(EthereumEvent.ACCOUNTS_CHANGED, onAccountsChanged);
-      internalProvider.on(EthereumEvent.CHAIN_CHANGED, onChainChangedOrDisconnect);
-      internalProvider.on(EthereumEvent.DISCONNECT, onChainChangedOrDisconnect);
+    if (
+      internalConnectedProvider &&
+      "on" in internalConnectedProvider &&
+      typeof internalConnectedProvider.on === "function"
+    ) {
+      internalConnectedProvider.on(EthereumEvent.ACCOUNTS_CHANGED, onAccountsChanged);
+      internalConnectedProvider.on(EthereumEvent.CHAIN_CHANGED, onChainChangedOrDisconnect);
+      internalConnectedProvider.on(EthereumEvent.DISCONNECT, onChainChangedOrDisconnect);
+    }
+
+    if (env) {
```

Comment:
I would move this block to a new `useEffect` hook on top of this one. I think that it would make it easier to read and each hook will do just one thing:

- One will set the readonly providers.
- The other one will attach and detach provider handlers.

Reply:
Makes sense. Done. I moved it below though, to avoid having the constants `connectProvider` and `disconnectProvider` far from where they are used.